Fauna-RPG/Entity/Components/HealthComponent.gd

```diff
@@ -0,0 +1,72 @@
+extends EntityComponent
+class_name HealthComponent
+## Holds player health values and allows for dealing damage.
+
+
+signal on_health_changed(new_health : int)
+
+@export var max_health : float = 100
+var current_health : float = 100
+
+
+func _ready() -> void:
+	_set_health(max_health)
+
+
+# Allow attacker to be null, but not the victim.
+# If attacker is null, we assume that the damage came from an unknown source,
+# so buffs won't calculate.
+# Use a negative damage value if you want to heal.
+func deal_damage(damage_data : DealDamageData) -> void:
+	var damage = damage_data.damage
+
+	if damage == 0.0:
+		return
+	
+	assert(owner != null, "No owner was set. Please call init on the Entity.")
+	
+	var victim : Entity = entity_owner
+	var attacker : Entity = damage_data.attacker
+	
+	# if this was a self attack, ignore the attacker
+	if attacker == victim: 
+		attacker = null
+	
+	# get stats from our stat components. We need to duplicate otherwise
+	# we're changing the base stats
+	var modified_attacker_stats : EntityStats = null
+	if attacker != null:
+		modified_attacker_stats = attacker.get_stat_component().stats.duplicate()
+	var modified_victim_stats : EntityStats = victim.get_stat_component().stats.duplicate()
+	
+	# apply buffs
+	if attacker != null and !damage_data.ignore_attacker_buffs:
+		for buff in attacker.get_buff_component().current_buffs:
+			buff.get_modified_stats(modified_attacker_stats)
```

Comment:
I don't think it really makes sense to apply the stats here. `get_modified_stats` is here for permanent stat changes. Thus at the start of each turn, when calling `apply_turn_start_buffs` it should apply every buff (whether that is a status effect like `Poison`, or a stat change like `Strength`). If there are status changes during the turn, we can check those status changes on the end of the turn of each entity / team
The `deal_damage` function should thus only do what is meant for : deal damages; and not change stats

Reply:
So just to be clear, you're saying you prefer option 2 where we need to revert stat changes when the buff is removed? In that case, I'm not sure we need to have "base" and "modified", because I don't see any cases where the base stats would be anything but 0. Unless we have different characters that start with different base stats, but even in that context I feel like it makes more sense to give that character a persistent buff. My intention with the stats was mainly just something for buffs/equipment to modify.

Reply:
Sounds good to me